File: csharp-project/RekindleBookStoreAqa/RekindleTaf/Model/Factory/CustomerFactory.cs
using System.Collections.Generic;
using Bogus;
using RekindleTaf.Exceptions;
using RekindleTaf.Utils;

namespace RekindleTaf.Model.Factory;

public class CustomerFactory : IFactory<Customer>
{
    private static readonly Faker<Customer> CustomerFaker = new Faker<Customer>()
        .RuleFor(customer => customer.FirstName, faker => faker.Person.FirstName)
        .RuleFor(customer => customer.LastName, faker => faker.Person.LastName)
        .RuleFor(customer => customer.Username, faker => faker.Person.UserName);

    private IDictionary<IValueKey, Supplier<Customer>>
        Map = new Dictionary<IValueKey, Supplier<Customer>>()
        {
            {CustomerKey.StdUser, () => CustomerFaker.Generate()}
        };

    public Customer Create(IValueKey key)
    {
        if (Map.TryGetValue(key, out Supplier<Customer>? value))
        {
            return value.Invoke();
        }

        throw new ValueObjectNotFoundException($"Customer identified by {key} has not yet been implemented");
    }
}

File: typescript-project/rekindle-book-store-aqa/src/model/customer-factory.ts
import {Customer} from "@/model/domain-model";
import {faker} from '@faker-js/faker';
import {Factory} from "@/model/factory";

export const CustomerKey = {
    stdCustomerActive: Symbol('StdCustomerActive')
};

export class CustomerFactory implements Factory<Customer> {

    private readonly map: Map<Symbol, Supplier<Customer>> = new Map([
        [CustomerKey.stdCustomerActive, () => this.createActiveCustomer()],
    ]);

    private createActiveCustomer(): Customer {
        return {
            firstName: faker.person.firstName(),
            lastName: faker.person.lastName(),
            username: faker.internet.userName()
        }
    }

    create(key: Symbol): Customer {
        const supplier = this.map.get(key);
        if (supplier) {
            return supplier();
        }
        throw new Error(`Bookstore identified by ${key.toString()} has not yet been implemented`);
    }

}